core/src/main/java/org/jdbi/v3/core/internal/CopyOnWriteHashMap.java

```diff
@@ -0,0 +1,325 @@
+/*
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.jdbi.v3.core.internal;
+
+import java.util.AbstractMap;
+import java.util.AbstractSet;
+import java.util.Collection;
+import java.util.Collections;
+import java.util.HashMap;
+import java.util.Iterator;
+import java.util.Map;
+import java.util.Objects;
+import java.util.Set;
+import java.util.Spliterator;
+import java.util.concurrent.atomic.AtomicReference;
+import java.util.function.BiConsumer;
+import java.util.function.BiFunction;
+import java.util.function.Function;
+
+public class CopyOnWriteHashMap<K, V> implements Map<K, V> {
+    private final AtomicReference<Map<K, V>> m;
+
+    public CopyOnWriteHashMap() {
+        m = new AtomicReference<>(new HashMap<>());
+    }
+
+    public CopyOnWriteHashMap(Map<K, V> that) {
+        if (that.getClass() == CopyOnWriteHashMap.class) {
+            m = new AtomicReference<>(((CopyOnWriteHashMap<K, V>) that).m.get());
+        } else {
+            m = new AtomicReference<>(new HashMap<>(that));
+        }
+    }
+
+    @Override
+    public int size() {
+        return m.get().size();
+    }
+
+    @Override
+    public boolean isEmpty() {
+        return m.get().isEmpty();
+    }
+
+    @Override
+    public boolean containsKey(Object key) {
+        return m.get().containsKey(key);
+    }
+
+    @Override
+    public boolean containsValue(Object value) {
+        return m.get().containsValue(value);
+    }
+
+    @Override
+    public V get(Object key) {
+        return m.get().get(key);
+    }
+
+    @Override
+    public V put(K key, V value) {
+        return m.getAndUpdate(old -> {
+            Map<K, V> updated = new HashMap<>(old);
+            updated.put(key, value);
+            return updated;
+        }).get(key);
+    }
+
+    @Override
+    public V putIfAbsent(K key, V value) {
+        return m.getAndUpdate(old -> {
+            if (old.get(key) != null) {
+                return old;
+            }
+            Map<K, V> updated = new HashMap<>(old);
+            updated.putIfAbsent(key, value);
```

Comment:
return value is ignored. How do you know if the map was updated ? You can't verify this with compile time checks, so here we return a copy on write even if the old map hasn't even changed.